WIP TypeScript refactor

kincaidoneil opened the pull request:
### Big Things

- Refactor in TypeScript a la xrp-asym-server
- Add balance logic; options for `maximum`, `settleTo` and `settleThreshold`
- Add channel watcher to claim channels if the sender tries to settle, if it's profitable
- Support for unidirectional sending: server will accumulate funds for an account up to the cost of an on-chain transaction, and only when it hits that threshold, open an outgoing payment channel (the client could either open a paychan to the server to front their cost, or keep fulfilling and wait until enough funds to it accumulated) [(More details on our outline/proposal for this here)](https://github.com/interledger/rfcs/issues/443).

### Little Things

- Fee estimation logic for different operations
- Add maxPacketAmount limiting
- Add support for `ilp-logger`
- Use store wrapper from xrp-asym-server
- Simplify account schema for now; use change detection to trigger updates to store
- Simplify BTP subprotocols to `info`: retrieves the Ethereum address the peer wants to be paid at; and `machinomy`: sends serialized Machinomy payments (also used to prove channel ownership)
- Prevent multiple simultaneous funding transactions
- Added upstream Machinomy fixes for payment validation and best claim
- Add account blocking
- Add max packet amount config
- Other miscellaneous fixes

### What's Next

- This has not been tested at all, so we expect a lot of bugs (we wanted to push this out as a work in progress first to solicit feedback). We're currently in the process of writing unit tests and pushing out fixes, but we'd love any feedback or help with tests!
- We're also looking to add black-box/integration tests as well so the whole stack have better security guarantees (https://github.com/machinomy/machinomy/issues/232)
- Open question for us: what is the optimal DoS prevention scheme? With the xrp plugins, the client has to escrow some multiple of what the server locks up (10x right now). In the case of Ethereum, we're not sure that's a sufficient mitigation due to higher fees, because if the server gets a payment for less than the cost to claim the channel, but still forwards the payment, the attacker can force the server to lose money (even though the attacker would likely lose more). Should a client with an incoming paychan have to prefund (to cover the server's fee if they needed to claim the channel) before the server begins crediting them any balance? Looking for ideas on the best approach

(Credit to @karzak for a ton of help with this!)

kincaidoneil committed:
Add TypeScript config and update deps

kincaidoneil committed:
Remove old test scripts

kincaidoneil committed:
Refactor plugin with TypeScript

kincaidoneil committed:
Add tslint config

Commit:
Unit Tests + BugFixes (#15)

* Begin eth asym server tests

Write index, file, create simple instance, create simple memory store for
tests.

* test skeleton

* currency scale check

* create plugin base done

* Default framework and packages installed

* test: unit tests

Add jest unit tests for main plugin. WIP -- still have not finished.
Create simple memstore for tests. Jest config in package.json to work
properly with node (had a bug when not explicitly specifying node
environment).

* fix: various fixes

Fee estimation: now returns in gwei instead of wei as intended.

Account store: assign anonymous function to proxy to fix `this`
context (was not setting to plugin store). Also, set the object property
value on the set trap (was not persisting account changes to store).

Still a WIP, various TODOs throughout.

* test: more unit tests

* fix: trace class + timer race functioning

Throwing traces fixed (previously not catching). Replace ternary
operator logic of _handlePrepareResponse which was not working before.
Promise race in _handleCustomData was outputting invalid structure for
ILP reject, now properly returns.

* chore: eslint and npm scripts

* fix: assorted bug fixes

* formatAmount fixed to handle negative numbers
* channel state should be of 0 not 1 for open channel
* added wei <> gwei conversion missing in balance logic
* timer adjusted from seconds to milliseconds

* test: all functions unit tested

* chore: install & configure eslint

* Update CircleCI config to run tests

kincaidoneil committed:
Refactor into single client/server plugin, replace Machinomy with Minomy

kincaidoneil committed:
Merge branch 'ko-refactor' into ko-refactor

kincaidoneil committed:
chore: upgrade ilp-packet

kincaidoneil committed:
fix: wrong entrypoint

kincaidoneil committed:
Add fixes, eth address sharing, receiveOnly mode

- fix: BigNumber exponential config
- feat: exchanging Ethereum address with peers
- fix: tx fee gwei/wei conversions
- fix: longer BTP timeout
- feat: add receiveOnly, fund outgoing channel by default on connect
- fix: no version property
- fix: trace level logging

kincaidoneil committed:
feat: add auto settlement and update balance middleware (part 1)

kincaidoneil committed:
feat: add settler (part 2)

kincaidoneil committed:
fix: settlement bugs

- fix: handle multiple settlement attempts simultaneously (safely) without triggering an infinite loop
- fix: only call settle (1) on client connect, (2) after FULFILL, (3) in the case of a simultaneous settle attempt, try again after settling is complete (e.g. if we received money)
- fix: correct claim increment logic
- fix: best outgoing claim not persisted
- fix: add and sub balance are now separate methods
- fix: client wrapper not passing data and money handler to plugin
- fix: client didn't resolve _connect promise
- fix: logging (sigh)
- feat: validate balance config in constructor

kincaidoneil committed:
feat: add naive claim on disconnect

kincaidoneil committed:
fix: miscellaneous

- feat: add config for outgoing channel amount
- fix: single Ethereum address for lifetime of account wasn't enforced
- feat: add commented-out channel watcher code
- fix: stricter validation of BTP subprotocol data
- fix: don't add/sub 0 from balance
- fix: don't advance to next settle method if no "budget" is leftover
- chore: better logging during settlement attempts
- fix: BigNumber thinks 0 is positive, accepting claims for 0 increments
- fix: credit from claims is capped at channel value
- chore: clarify settlement variable names

kincaidoneil committed:
docs: add WIP readme

kincaidoneil committed:
chore: reorg source files, remove minomy dep

kincaidoneil committed:
fix: queue incoming claims to prevent race condition

Validating incoming claims requires async operations, which caused non-deterministic errors when multiple claims were handled within a short time interval, such as:
- moneyHandler was invoked with incorrect amounts
- credited wrong claim increment to balance
- 2nd claim would sometimes be invalid because the 1st claim hadn't yet been persisted

kincaidoneil committed:
feat: add channel watcher, initiation fee, queue outgoing settlements, other fixes

- feat: add config for fee for incoming channels
- feat: queue outgoing settlements, with limit of 1 in queue
- fix: save payout amount to limit outgoing settlements to the total packets fulfilled, preventing a connector from just sending incoming money directly back to the sender (if not prefunding)
- feat: add channel watcher and claim if profitable functionality
- feat: claim on disconnect, by default on client
- fix: checking if a channel is already linked to account is safer and more robust
- fix: add unique key appendices for store
- fix: cleanup on disconnect, wait for settlements to complete, and unload store
- fix: remove unnecessary isBlocked functionality (for now)
- chore: rename subprotocol to 'machinomy'
- fix: parse balance as BigNumber when loading store
- feat: add settleOnConnect and claimOnDisconnect options

kincaidoneil committed:
fix: add 'peer' role to plugin options

kincaidoneil committed:
fix: await settle on connect

kincaidoneil committed:
chore: add more logs to settlement and elsewhere

kincaidoneil committed:
fix: settleTo >= settleThreshold

- settlement is triggered when the balance is less than the settleThreshold
- settleTo can equal the threshold, since there should always be a positive, non zero amount to settle
- this is more inline with the connector balance middleware

kincaidoneil committed:
fix: web3 BigNumber issues

kincaidoneil committed:
feat: add private key, provider config options

- Abstracts Web3 instantiation and installation so consumers of plugin don't have to deal with it
- Also should fix moneyd issues per: https://github.com/interledgerjs/ilp-plugin-ethereum-asym-server/pull/16

kincaidoneil committed:
fix: fork of btp/mini-accounts

kincaidoneil committed:
fix: prefix private key with 0x

kincaidoneil committed:
docs: add api updates and balance explanation

kincaidoneil committed:
docs: settleTo explanation clarity

kincaidoneil committed:
chore: rename module, update package.json

- remove ganache-cli dep
- update description
- add keywords and other package.json fields
- update Circle config with new module name for releasing NPM plugin
- pin web3 to specific version

kincaidoneil committed:
fix: validate eth private key

Also, it works with or without 0x in front!

kincaidoneil committed:
fix: ilp-logger in typescript

kincaidoneil committed:
fix: mainnet contract ABI

- @machinomy/contracts 4.1.1 had the wrong ABI for the mainnet contract
- Remove it as a dep, and just include the ABI files here directly

kincaidoneil committed:
chore: cleanup packages

- remove @machinomy/contracts
- remove eslint
- use the plugin-btp updates!
- temporarily remove Dino's unit tests

kincaidoneil committed:
fix: import for json ABI files

kincaidoneil committed:
test: add initial integration test

- Use Ropsten since ganache-cli had issues with gas calculation
- Definitely needs to be restructured and more robust tests (this is more of an MVP)

kincaidoneil committed:
fix: mini-accounts version

kincaidoneil committed:
fix: default options

- fix: resolve bug where max balance would default to infinity even though it was configured with 0 (due to the falsy value)
- fix: only pass non-Ethereum plugin specific options to plugin-btp/mini-accounts
- much cleaner defaults using ES6 destructuring
- default min incoming settlement period is now 3 days
- default outgoing is now 6 days

kincaidoneil committed:
feat: event-based channel watcher, prevent simultaneous claim txs

- subscribes to push notifications for contract settling
- should attempt to claim the channel instantly if the peer starts settling

kincaidoneil committed:
test: fix issues, use ava

- due to https://github.com/facebook/jest/issues/3917, Jest assertions weren't working from within Stream event handlers
- miscellaneous fixes

kincaidoneil committed:
test: channel watcher

- temporarily remove event-based watcher due to web3 bug
- add new watcher interval configuration
- minor fix to initializer in constructor

kincaidoneil committed:
chore: run builds sequentially

- Since the Node 8 & Node 10 tests are using the same Ethereum accounts, one will fail if run in parallel
- Previously triggered "underpriced replacement tx" error

kincaidoneil committed:
style: fix lint errors

kincaidoneil committed:
chore: prepare for publishing

- update Circle config with publishing to NPM @next tag
- add code coverage with nyc
- only include built files in NPM package
- simplify NPM scripts
- add badges to README

kincaidoneil committed:
chore: fix circle

- incorrect regex for publishing new modules based on tags

kincaidoneil committed:
fix: circle not publishing

kincaidoneil committed:
chore: better source maps

kincaidoneil committed:
test: web3 provider as environment variable

- include the memory store locally

kincaidoneil committed:
fix: npm home dir on circle

kincaidoneil committed:
3.0.0-beta.2

kincaidoneil committed:
feat: settle on T04 error

- a la https://github.com/interledgerjs/ilp-plugin-xrp-asym-server/pull/54
- essentially, resolves deadlocks if a settlement failed previously
- e.g. plugin successfully opens a channel, then tries but fails to sign a claim because the transaction hasn't propagated yet

kincaidoneil committed:
3.0.0-beta.3

kincaidoneil committed:
chore: simplify circle config and publishing

kincaidoneil committed:
chore: cleanup comments, fix types

emschwartz commented:
For the future, please submit more smaller PRs. It's nearly impossible to do a proper review of a 50+ commit PR

emschwartz commented on src/__tests__/stream.test.ts:
```diff
@@ -0,0 +1,148 @@
+  let serverStream: IlpStream.DataAndMoneyStream
+
+  let clientConn: IlpStream.Connection
+  let clientStream: IlpStream.DataAndMoneyStream
```
Why are these declared above where they're used?

emschwartz commented on src/account.ts:
```diff
@@ -0,0 +1,843 @@
+   *      return the original amount
+   *    - e.g. if it sent a claim for the entire amountToSettle, it will return 0 because there's
+   *      nothing leftover
+   * - This is really cool, because the amount to settle can "flow" from one settlement method to another,
```
What are the different settlement methods you're referring to here?

kincaidoneil replied:
`sendClaim` and `fundOutgoingChannel`

emschwartz commented on src/account.ts:
```diff
@@ -0,0 +1,843 @@
+        // 3) Try to send a claim again since we may have more funds in the channel
+        if (amountLeftover.lte(0)) throw amountLeftover
+        amountLeftover = await this.sendClaim(amountLeftover)
+      } catch (leftover) {
```
It's strange to throw things that aren't errors. What if `this.sendClaim` actually throws an error? Seems like amountLeftover would be wrong

emschwartz commented on src/account.ts:
```diff
@@ -0,0 +1,843 @@
+
+  async handleData (message: BtpPacket, dataHandler?: DataHandler): Promise<BtpSubProtocol[]> {
+    const info = getSubprotocol(message, 'info')
+    const ilp = getSubprotocol(message, 'ilp')
```
nit: don't need to check for both if it's only valid to have included one of them (and one will be ignored if you set both anyway)

kincaidoneil replied:
Yeah, it'd be weird for the packet to include both. I think having both checks is useful though in case they sent some random subprotocol name that we don't support.

emschwartz commented on src/account.ts:
```diff
@@ -0,0 +1,843 @@
+          }
+
+          // Always fetch updated channel state from network
+          const channel = await fetchChannel(this.master._web3, claim.channelId)
```
Does it really make sense to make that API call every time you get a claim in? If the channels are unidirectional you shouldn't need to do that unless you think they might have put more money in the channel and you want to check if they did

kincaidoneil replied:
~~Maybe not, but it might be a premature optimization. I think it's safer to do it each claim. For example, if the channel was claimed outside of the plugin and the state was not fetched, new claims would still be credited.~~

kincaidoneil replied:
Disregard. You're right. In my testing, processing a claim took 300-500ms even with a 4ms latency connection, which--given the incoming claims are queued--is unacceptable if settlement is triggered on every packet. Furthermore, even if no funding tx is required, the channel state is also queried before signing an outgoing claim.

I'll add caching of the channel state for both incoming and outgoing channels. This assumes that a user doesn't deposit to or claim their channel outside of the plugin, but that's probably a reasonable assumption.

emschwartz commented on src/account.ts:
```diff
@@ -0,0 +1,843 @@
+            throw new Error(`signature is invalid, or sender is using contracts on a different network`)
+          }
+
+          if (new BigNumber(claim.value).lte(0)) {
```
nit: check this before checking the signature

emschwartz commented on src/account.ts:
```diff
@@ -0,0 +1,843 @@
+            this.master._store.set(channelKey, this.account.accountName)
+            this.master._log.trace(`Incoming channel ${claim.channelId} is now linked to account ${this.account.accountName}`)
+
+            // TODO web3.js beta 36 event bugs: https://github.com/ethereum/web3.js/issues/1916
```
This seems kind of important, no?

kincaidoneil replied:
The channel watcher polls at regular intervals, so it should still work fine. The event subscriptions are just nice because it'd trigger it to claim the channel instantly (if it was profitable). I'll add it back in after the Web3 bug is fixed.

emschwartz commented on src/utils/queue.ts:
```diff
@@ -0,0 +1,45 @@
+import 'source-map-support/register'
```
Apparently there are arguments against including this in libraries, cause it doesn't play nicely with webpacking things

emschwartz commented on src/__tests__/stream.test.ts:
```diff
@@ -0,0 +1,148 @@
+    setTimeout(resolve, 10000)
+  })
+
+  t.true(actualReceived.gte(AMOUNT_TO_SEND.minus(INCOMING_FEE)), 'server received at least as much money as the client sent')
```
Would this fail if the network fees increase?

kincaidoneil replied:
That test should still work. `INCOMING_FEE` is a fixed initiation fee that the connector charges when it gets a claim from a new channel (to cover the cost of them claiming it, although right now it's not dynamic to the cost of the claim tx).

However, the tests _would_ fail if the fee to open a channel was greater than the `settleTo` amount in the sender's balance configuration (since even on the client, the fee to open a channel is added to/debited from the balance). That said, the prefund amount is `0.002` eth, so the transaction fees should almost always be less than that.

kincaidoneil committed:
chore: remove source-map-support

sentientwaffle commented on src/account.ts:
```diff
@@ -0,0 +1,843 @@
+
+      return settlementBudget
+    } catch (err) {
+      this.master._log.error(`Failed to fund outgoing channel to ${this.account.accountName}: ${err.message}`)
```
Since the `settlementBudget` is sometimes thrown, this error doesn't make sense (and `settlementBudget.message` is probably `undefined`). As evan mentioned, throwing amounts is hard to follow. If there are different amounts/states that need to be returned, maybe return an object with one of 2+ fields.

kincaidoneil replied:
I updated it so it only rejects the promise in those cases rather than throwing. Is there a cleaner way to do this?

sentientwaffle replied:
Hm, I see. That resolves the error message issue, but I'm still concerned about rejecting with non-error objects. It's a valid but unexpected pattern, which leads to subtle errors. For example:

```javascript
      } catch (leftover) {
        // If no money was sent, rejections should return the original amount
        if (BigNumber.isBigNumber(leftover)) {
          amountLeftover = leftover
        }
      }
```

If the caught error isn't a `BigNumber`, it is discarded -- is that the correct behavior?

For clarity, I'd suggest having the function return `{amountLeftover, success: true|false}` and never reject with an amount. Instead, the caller would check `success` and behave accordingly. It's a little more verbose, but I think it's easier to follow.

kincaidoneil committed:
fix: clarity in tests

sentientwaffle commented on src/plugin.ts:
```diff
@@ -0,0 +1,117 @@
+    responsePacket: IlpPacket.IlpPacket,
+    preparePacket: IlpPacket.IlpPacket
+  ): Promise<void> =>
+    this._getAccount(destination).handlePrepareResponse(preparePacket, responsePacket)
```
`handlePrepareResponse` doesn't return a `Promise` -- this only type checks because `_handlePrepareResponse` is marked `async`. Also, why use `_handlePrepareResponse = async (...)` instead of `async _handlePrepareResponse (...`.

kincaidoneil replied:
Fixed the promise ✅

When I define it that way, TypeScript throws an error that it's an instance member property and not an instance member function (I think it has to do with this: https://github.com/interledgerjs/ilp-plugin-mini-accounts/blob/master/src/index.ts#L65).

sentientwaffle commented on src/__tests__/watcher.test.ts:
```diff
@@ -0,0 +1,81 @@
+
+      if (wasClaimed) {
+        t.pass()
+        resolve()
```
`clearInterval()` so that this doesn't keep running after the promise is resolved.

kincaidoneil committed:
fix: address comments

kincaidoneil committed:
chore: TS to 3.1.1, fix README

kincaidoneil committed:
chore: add codecov to CI

kincaidoneil committed:
fix: don't throw non-errors

- use async reducer for "flowing" budget between settlement methods

kincaidoneil committed:
fix: ts error in tests

kincaidoneil committed:
fix: linting

kincaidoneil committed:
fix: cleanup settlement logic

kincaidoneil committed:
Create LICENSE

sentientwaffle commented on src/account.ts:
```diff
@@ -413,18 +418,16 @@ export default class EthereumAccount {
       this.master._log.error(`Failed to fund outgoing channel to ${this.account.accountName}: ${err.message}`)
-      throw settlementBudget
+    } finally {
+      return settlementBudget
```
nit: I think it would be clearer if this `return` was in the `catch` with the log line, since if no error is thrown it will return from within the `try` block.

kincaidoneil replied:
Are you sure? Maybe I'm misinterpreting, but I think there are a number of cases where it wouldn't return the settlement budget from within the try (unless I added a return within both the try and catch).

kincaidoneil committed:
chore: fix codecov, package.json

kincaidoneil committed:
test: deposit tx, smaller streaming chunks

- update package-lock
- remove `html` reporter from nyc

kincaidoneil committed:
test: increase stream timeout

kincaidoneil committed:
chore: update ava, mini-accounts

kincaidoneil committed:
feat: cache channel state when signing/validating claims

- resolves streaming issues due to fetching network channel state while on high-latency connections
- note: this needs further review for security concerns before used in production
- wait for 1 confirmation after opening channel -- may fix issue where receipt is available, but channel is updated when fetching state from the network moments later

kincaidoneil committed:
3.0.0-beta.4

kincaidoneil committed:
fix: circleci not building

kincaidoneil committed:
3.0.0-beta.5

kincaidoneil committed:
fix: simplify settlement logic

kincaidoneil commented:
@sentientwaffle @emschwartz Is this good to merge? Still intend to only release as betas, but I think it's stable enough for master.

kincaidoneil committed:
fix: store and logging per LPIv2

- fix store not being created properly when used in a connector
- use LPIv2 services parameter rather than _store and _log
- default to memory store if no store was provided
- pass store and log to internal plugin

kincaidoneil committed:
fix: add delay before sending claim

- wait 2s for block to propagate before sending a claim
- sometimes the channel lookup would fail if we do it too soon

kincaidoneil committed:
3.0.0-beta.6

kincaidoneil committed:
fix: reference to data and money handlers

kincaidoneil committed:
fix: strict mode in tsconfig

kincaidoneil committed:
fix: default and named exports

- TS export = syntax for commonjs compat doesn't support named exports
- Now the module is commonjs compatible with default and named exports

kincaidoneil committed:
temp: disable strict in TS, pending plugin-btp updates

- TS tries to recompile ilp-plugin-btp, which fails strict
- interledgerjs/ilp-plugin-btp#45 should fix

kincaidoneil committed:
fix: linting error

kincaidoneil committed:
3.0.0-beta.7

kincaidoneil committed:
chore: update README
